scipy/sparse/construct.py

```diff
@@ -622,6 +622,85 @@ def bmat(blocks, format=None, dtype=None):
 
     return coo_matrix((data, (row, col)), shape=shape).asformat(format)
 
+def _block_diag(mats, format=None, dtype=None):
+    """      
+    Build a block diagonal sparse matrix from equally shaped matrices.
+    Much faster than block_diag for big matrices
+
+    Parameters
+    ----------
+    mats : sequence of equally shaped matrices
+        Input matrices.
+    format : str, optional
+        The sparse format of the result (e.g. "csr").  If not given, the matrix
+        is returned in "coo" format.
+    dtype : dtype specifier, optional
+        The data-type of the output matrix.  If not given, the dtype is
+        determined from that of `blocks`.
+
+    Returns
+    -------
+    res : sparse matrix
+
+    Notes
+    -----
+
+    .. versionadded:: 0.18.0
+
+    See Also
+    --------
+    bmat, diags, block_diag
+
+    Examples
+    --------
+    >>> from scipy.sparse import coo_matrix, block_diag
+    >>> A = coo_matrix([[1, 2], [3, 4]])
+    >>> B = coo_matrix([[5, 6], [7, 8]])
+    >>> C = coo_matrix([[9, 10], [11,12]])
+    >>> _block_diag((A, B, C)).toarray()
+    array([[ 1,  2,  0,  0,  0,  0],
+           [ 3,  4,  0,  0,  0,  0],
+           [ 0,  0,  5,  6,  0,  0],
+           [ 0,  0,  7,  8,  0,  0],
+           [ 0,  0,  0,  0,  9, 10],
+           [ 0,  0,  0,  0, 11, 12]])
+    """
+    
+    n = len(mats)
+    shape = mats[0].shape
+
+    if any(mat.shape != mats[-1].shape for mat in mats) or (
+            any(issparse(mat) for mat in mats)):
+        data = []
+        col = []
+        row = []
+        origin = np.array([0, 0], dtype=np.int)
+        for mat in mats:
+            if issparse(mat):
+                data.append(mat.data)
+                row.append(mat.row + origin[0])
```

Comment:
The input matrices aren't necessarily in coo format, need to do `mat = mat.tocoo()`